Druid SQL API documentation refactor (#14711)
apache/druid: docs/api-reference/sql-api.md

## Query from Historicals

### Submit a query

Submit a SQL-based query in the JSON request body. Returns a JSON object with the database results and a set of header metadata associated with the query. 

This endpoint also supports querying metadata by using [metadata tables](../querying/sql-metadata-tables.md).

#### URL

<code class="postAPI">POST</code> <code>/druid/v2/sql</code>

#### Request body

* `query`: SQL query string.
* `resultFormat`: Format of query results.
  * `object`: Returns a JSON array of JSON objects with the HTTP header `Content-Type: application/json`.
  * `array`: Returns a JSON array of JSON arrays with the HTTP header `Content-Type: application/json`.
  * `objectLines`: Returns newline-delimited JSON objects with a trailing blank line. Sent with the HTTP header `Content-Type: text/plain`.
  * `arrayLines`: Returns newline-delimited JSON arrays with a trailing blank line. Sent with the HTTP header `Content-Type: text/plain`. 
  * `csv`: Returns a comma-separated values with one row per line and a trailing blank line. Sent with the HTTP header `Content-Type: text/csv`. 

> * `header`: Adds a header row with information on column names for the query result when set to `true`. You can optionally include `typesHeader` and `sqlTypesHeader`. 

Reviewer: What do you think about including example snippets for when these Boolean values are set to true? Something like:

### for the header description
`````
When `header` is `true`, Druid includes a row similar to the following:
* For an `object` or `objectLines` result format:
   ```json
   <EXAMPLE>
   ```

* For an `array`, `arrayLines`, or `csv` result format:
   ```json
   <EXAMPLE>
   ```
`````

### for the typesHeader description
`````
When `header` is `true` and `typesHeader` is true, Druid includes a row similar to the following:
* For an `object` or `objectLines` result format:
   ```json
   <EXAMPLE>
   ```

* For an `array`, `arrayLines`, or `csv` result format:
   ```json
   <EXAMPLE>
   ```
`````

And then also something similar for sqlTypesHeader.

That way we can capture some of the information from the results format table:
> If header is true, the first row is an object where the fields are column names. Each field's value is either null (if typesHeader and sqlTypesHeader are false) or an object that contains the Druid type as type (if typesHeader is true) and the SQL type as sqlType (if sqlTypesHeader is true).

Reviewer: Though this can get unwieldy. Consider other ways to format and present this information too.

    Complex types, like sketches, will be reported as `COMPLEX<typeName>` if a particular complex type name is known for that field, or as `COMPLEX` if the particular type name is unknown or mixed. 
    

>     If working with an older version of Druid, set `header` to `true` to verify compatibility. Druid returns the HTTP header `X-Druid-SQL-Header-Included: yes` when the client connects to an older Druid version with support for the `typesHeader` and `sqlTypesHeader` parameters. Additionally, Druid returns a `X-Druid-SQL-Query-Id` HTTP header with the value of `sqlQueryId` from the [query context parameters](../querying/sql-query-context.md) if specified, else Druid will generate a SQL query ID.

Reviewer: ```suggestion
    Older versions of Druid that support  the `typesHeader` and `sqlTypesHeader` parameters return the HTTP header `X-Druid-SQL-Header-Included: yes` when you set `header` to `true`. Druid returns the HTTP response header for compatibility, regardless of whether `typesHeader` and `sqlTypesHeader` are set.
```
Suggested rewording to retain the original message

* `typesHeader`: Adds Druid runtime type information in the header. Can only be set when `header` is also `true`.
* `sqlTypesHeader`: Adds SQL type information in the header. Can only be set when `header` is also `true`.
* `context`: JSON object containing optional [SQL query context parameters](../querying/sql-query-context.md), such as to set the query ID, time zone, and whether to use an approximation algorithm for distinct count.

> * `parameters`: List of query parameters for parameterized queries. Each parameter in the list should be a JSON object like `{"type": "VARCHAR", "value": "foo"}`. The type should be a SQL type; see [Data types](../querying/sql-data-types.md) for a list of supported SQL types.

Reviewer: `````suggestion
* `parameters`: List of query parameters for parameterized queries. Each element in the array should be a JSON object containing the parameter's SQL data type and the parameter value. For a list of supported SQL types, see [Data types](../querying/sql-data-types.md).
`````

> 

Reviewer: Making the example more prominent with the property name will be easier to follow. Another option that could help users is to include the parameterized query as a separate example, similar to what was in the docs
![image](https://github.com/apache/druid/assets/7747997/9e952721-637c-4f44-9adb-9aadfd94c1ab)

`````suggestion
For example:
   ```
    "parameters": [
        {
            "type": "TIMESTAMP",
            "value": "2019-08-21 00:00:00"
        }
    ]
   ```
`````

#### Responses

<!--DOCUSAURUS_CODE_TABS-->

<!--200 SUCCESS-->

*Successfully submitted query* 

<!--400 BAD REQUEST-->

*Error thrown due to bad query. Returns a JSON object detailing the error with the following format:* 

```json
{
    "error": "A well-defined error code.",
    "errorMessage": "A message with additional details about the error.",
    "errorClass": "Class of exception that caused this error.",
    "host": "The host on which the error occurred."
}
```
<!--500 INTERNAL SERVER ERROR-->

*Request not sent due to unexpected conditions. Returns a JSON object detailing the error with the following format:* 

```json
{
    "error": "A well-defined error code.",
    "errorMessage": "A message with additional details about the error.",
    "errorClass": "Class of exception that caused this error.",
    "host": "The host on which the error occurred."
}
```

<!--END_DOCUSAURUS_CODE_TABS-->

---

#### Sample request

> The following example retrieves all rows in the `wikipedia` datasource where the `user` is `BlueMoon2662`. The query is assigned the ID `request01` using the `sqlQueryId` parameter. Optional properties `header`, `typesHeader`, and `sqlTypesHeader` are included to add type information to the response.

Reviewer: ```suggestion
The following example retrieves all rows in the `wikipedia` datasource where the `user` is `BlueMoon2662`. The query is assigned the ID `request01` using the `sqlQueryId` context parameter. The optional properties `header`, `typesHeader`, and `sqlTypesHeader` are set to true to include type information to the response.
```

        },
        "regionIsoCode": {
            "type": "STRING",
            "sqlType": "VARCHAR"
        },
        "regionName": {
            "type": "STRING",
            "sqlType": "VARCHAR"
        },
        "user": {
            "type": "STRING",
            "sqlType": "VARCHAR"
        },
        "delta": {
            "type": "LONG",
            "sqlType": "BIGINT"
        },
        "added": {
            "type": "LONG",
            "sqlType": "BIGINT"
        },
        "deleted": {
            "type": "LONG",
            "sqlType": "BIGINT"
        }
    },
    {
        "__time": "2015-09-12T00:47:53.259Z",
        "channel": "#ja.wikipedia",
        "cityName": "",
        "comment": "/* 対戦通算成績と得失点 */",
        "countryIsoCode": "",
        "countryName": "",
        "isAnonymous": 0,
        "isMinor": 1,
        "isNew": 0,
        "isRobot": 0,
        "isUnpatrolled": 0,
        "metroCode": 0,
        "namespace": "Main",
        "page": "アルビレックス新潟の年度別成績一覧",
        "regionIsoCode": "",
        "regionName": "",
        "user": "BlueMoon2662",
        "delta": 14,
        "added": 14,
        "deleted": 0
    }
]
```
</details>

### Cancel a query

> Cancels a query on the Router or the Broker with the associated `sqlQueryId`. Queries can only be canceled with a valid `sqlQueryId`. The `sqlQueryId` must be set in the query context when the query is submitted. Note that Druid does not enforce unique `sqlQueryId` in the query context. Druid cancels all requests that use the same query ID.

Reviewer: ```suggestion
Cancels a query on the Router or the Broker with the associated `sqlQueryId`. Queries can only be canceled with a valid `sqlQueryId`. The `sqlQueryId` must be set in the query context when the query is submitted. Note that Druid does not enforce a unique `sqlQueryId` in the query context. Druid cancels all requests that use the same query ID.
```

> Cancels a query on the Router or the Broker with the associated `sqlQueryId`. Queries can only be canceled with a valid `sqlQueryId`. The `sqlQueryId` must be set in the query context when the query is submitted. Note that Druid does not enforce unique `sqlQueryId` in the query context. Druid cancels all requests that use the same query ID.

Reviewer: Regarding this line:
>The `sqlQueryId` must be set in the query context when the query is submitted.

Technically it doesn't have to be set, right? The original docs state,
>Druid returns the SQL query identifier in the X-Druid-SQL-Query-Id HTTP header. This query id will be assigned the value of sqlQueryId from the [query context parameters](https://druid.apache.org/docs/latest/querying/sql-query-context.html) if specified, else Druid will generate a SQL query id for you.

So if you didn't send it with the query, you could still get it from the HTTP header to delete later.

When you cancel a query, Druid handles the cancellation in a best-effort manner. Druid immediately marks the query as canceled and aborts the query execution as soon as possible. However, the query may continue running for a short time after you make the cancellation request.

Cancellation requests require READ permission on all resources used in the SQL query. 

#### URL

<code class="deleteAPI">DELETE</code> <code>/druid/v2/sql/:sqlQueryId</code>

#### Responses

<!--DOCUSAURUS_CODE_TABS-->

<!--202 SUCCESS-->

*Successfully deleted query*

<!--403 FORBIDDEN-->

*Authorization failure* 

<!--404 NOT FOUND-->

*Invalid `sqlQueryId` or query was completed before cancellation request* 

<!--END_DOCUSAURUS_CODE_TABS-->

---

#### Sample request

> The following example cancels a request with specified query ID `request01`.

Reviewer: ```suggestion
The following example cancels a request with the query ID `request01`.
```

Cancellation requests require READ permission on all resources used in the SQL query. 

#### URL

<code class="deleteAPI">DELETE</code> <code>/druid/v2/sql/:sqlQueryId</code>

#### Responses

<!--DOCUSAURUS_CODE_TABS-->

<!--202 SUCCESS-->

*Successfully deleted query*

<!--403 FORBIDDEN-->

*Authorization failure* 

<!--404 NOT FOUND-->

*Invalid `sqlQueryId` or query was completed before cancellation request* 

<!--END_DOCUSAURUS_CODE_TABS-->

---

#### Sample request

The following example cancels a request with the set query ID `request01`.

<!--DOCUSAURUS_CODE_TABS-->

<!--cURL-->

```shell
curl --request DELETE "http://ROUTER_IP:ROUTER_PORT/druid/v2/sql/request01"
```

<!--HTTP-->

```HTTP
DELETE /druid/v2/sql/request01 HTTP/1.1
Host: http://ROUTER_IP:ROUTER_PORT
```

<!--END_DOCUSAURUS_CODE_TABS-->

#### Sample response

A successful response results in an `HTTP 202` message code and an empty response body.

### Query output format

> The following table is a comparative overview of different data formats, header configurations, and corresponding outputs.
> 
> ```
> | Format | typesHeader | sqlTypesHeader | Output                                                                             |
> |--------|-------------|----------------|------------------------------------------------------------------------------------|

Reviewer: ```suggestion
The following table shows examples of how Druid returns the column names and data types based on the result format and the type request.

| Format | typesHeader | sqlTypesHeader | Example output                                                                     |
|--------|-------------|----------------|------------------------------------------------------------------------------------|
```
